File: Arrays/Buy and Sell Stocks/code.py
#buy and sell stock
#Stock Buy Sell to Maximize Profit using Valley Peak Approach:

def maxProfit(price):
  maxprofit = 0
  
  profit = 0 #for profit from combination of days
  first_index = 0
  last_index  = 0


  for i in range(1,len(price)):
    #check the difference between between current price and previous price
    if price[i] > price[i-1]:
      maxprofit += price[i]- price[i-1] #stored in maxprofit
      
      profit    += price[i]- price[i-1] #store in profit these both ration
      last_index = i #make last_index as current index value

      if last_index == len(price)-1:
        print('start day: {}  End day: {} profit: {} '.format(first_index,last_index,profit))
    else:
      if last_index !=0: #if last index is not equal to 0 than print all the data
        print('start day: {}  End day: {} profit: {} '.format(first_index,last_index,profit))
      
      first_index = i #make first index as current i value
      last_index = 0 #last index again to 0 
      profit     = 0 #make profit is 0 for next price details

  return maxprofit



price = [100, 180, 260, 310, 40, 30, 535, 695]
# price = [100, 180, 260, 310, 40, 535, 695]
# price = []
maxProfit(price)

File: Arrays/Buy and Sell Stocks/code.java
// Java program for the above approach
// import java.io.*;

class GFG {

	static int maxProfit(int prices[], int size)
	{

		// maxProfit adds up the difference between
		// adjacent elements if they are in increasing order
		int maxProfit = 0, MS = 0, ML = -1, profit = 0;

		// The loop starts from 1
		// as its comparing with the previous
		for (int i = 1; i < size; i++) {
			if (prices[i] > prices[i - 1]) {
				maxProfit += prices[i] - prices[i - 1];
              	profit += prices[i] - prices[i - 1];
              	ML = i;
              
              if(i == size - 1) {
                System.out.println("Buy: " + MS + " Sell: " + ML + " Profit: " + profit); 
              }
            } else {
              if(ML != -1) {
              	System.out.println("Buy: " + MS + " Sell: " + ML + " Profit: " + profit);
              } 
              MS = i;
              ML = -1;
              profit = 0;
            }
        }
		return maxProfit;
	}

	// Driver code
	public static void main(String[] args)
	{

		// stock prices on consecutive days
		int price[] = { 100, 180, 260, 310, 40, 30, 535, 695 };
		int n = price.length;

		// function call
		System.out.println(maxProfit(price, n));
	}
}
